torch/csrc/jit/python_compiled_function.cpp

```diff
@@ -0,0 +1,219 @@
+#include "python_compiled_function.h"
+
+#include "torch/csrc/jit/pybind.h"
+#include "torch/csrc/autograd/variable.h"
+#include "torch/csrc/autograd/functions/jit_closure.h"
+#include "torch/csrc/jit/tracer.h"
+#include "torch/csrc/jit/passes/dead_code_elimination.h"
+#include "torch/csrc/jit/passes/peephole.h"
+#include "torch/csrc/jit/passes/graph_fuser.h"
+#include "torch/csrc/jit/passes/inplace_check.h"
+#include "torch/csrc/jit/python_arg_flatten.h"
+
+#include <algorithm>
+#include <functional>
+#include <atomic>
+
+namespace torch { namespace jit { namespace python {
+
+using namespace torch::autograd;
+using namespace torch::jit::tracer;
+
+namespace {
+
+// pybind casts are really verobse...
+py::object steal(py::handle x) {
+  return py::reinterpret_steal<py::object>(x);
+}
+
+py::object borrow(py::handle x) {
+  return py::reinterpret_borrow<py::object>(x);
+}
+
+} // anonymous namespace
+
+// Lifecycle of a CompiledFunction:
+//
+// - It is given an underlying function, which knows how to actually
+//   execute the code that we want to compile.
+// - When we encounter an input configuration for which we don't
+//   have an optimized trace, we run the underlying function, tracing its
+//   result.  The trace is not done yet, so we save it into our set of pending
+//   traces for that configuration.
+// - When we encounter an input configuration whose trace is "ready"
+//   (that is, we've seen all of the passes, so the trace contains
+//   forwards/backwards/etc), we compile it, and then register this
+//   as the compiled trace.
+// - When we encounter an input configuration whose trace is compiled,
+//   we just directly run the compiled trace.
+struct CompiledFunction {
+
+  struct TraceForKey {
+    TraceForKey(CompiledFunction& fn, bool is_volatile)
+      : fn_(fn)
+      , is_volatile_(is_volatile) {}
+
+    bool ready() {
+      if (closure_) return true;
+
+      // Remove expired traces
+      traces_.erase(std::remove_if(traces_.begin(),
+                                   traces_.end(),
+                                   [](const std::shared_ptr<TracingState>& state) {
+                                     return state->is_expired();
+                                   }),
+                    traces_.end());
+
+      // Check if any trace is complete
+      auto complete_it = std::find_if(traces_.begin(),
+                                   traces_.end(),
+                                   [](const std::shared_ptr<TracingState>& state) {
+                                     return state->is_complete();
+                                   });
+      if (complete_it == traces_.end())
+        return false;
+
+      auto complete_trace = *complete_it; // NOTE: copy, because we clear right after
+      traces_.clear();
+
+      // Now, we have a complete trace. Compile it.
+      EliminateDeadCode(complete_trace->graph);
```

Comment:
Not for this patch, but now that the optimizations live in C++, it would be really good to have a gcc/clang style way of toggling specific optimizations on/off from Python.

Reply:
Can you elaborate? I'm not sure what you meant by that